### ISSUE TITLE
Unmatched groups: sub vs subf

### ISSUE THREAD

**Issue by mrabarnett** [mrabarnett/mrab-regex#439]
**[Original report](https://bitbucket.org/mrabarnett/mrab-regex/issue/435) by Isaac Muse (Bitbucket: [facelessuser](https://bitbucket.org/facelessuser), GitHub: [facelessuser](https://github.com/facelessuser)).**

----------------------------------------

I understand roughly why things are behaving as they are, but with that said I'm curious if this was intentional or unintentional. Consider the following scenario with both `sub` and `subf`. In both cases we have two groups, only one of which "captures" anything.

```py
>>> regex.sub(r'(test1)|(test2)', r'matched: \1\2', 'test1')
'matched: test1'
>>> regex.subf(r'(test1)|(test2)', r'matched: {1}{2}', 'test1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/regex/regex.py", line 288, in subf
    return pat.subf(format, string, count, pos, endpos, concurrent, timeout)
TypeError: __str__ returned non-string (type NoneType)
```

Using the normal `sub` method, we will get an empty string in the replacement template for `\2` if nothing is captured. This matches behavior in Re as well.

When using `subf`, we run into an issue where the replacement just fails. Now, I understand that there are no captures, and you cannot index to the first capture as it does not exist, so logically this makes sense, but should it not behave at least similar to the non-format case when no indexing occurs? If there are no captures for an existing group, and the user has not indexed anything beyond the first capture, should it just return an empty string?

I also guess that there should be no expectation that `subf` should match `sub` in behavior, so I'm curious about your thoughts on this.

**Comment by mrabarnett** [comment]
**Original comment by Matthew Barnett (Bitbucket: [mrabarnett](https://bitbucket.org/mrabarnett), GitHub: [mrabarnett](https://github.com/mrabarnett)).**

----------------------------------------

It’s unintentional. I’ll fix it. Also unintentional was leaving this issue tracker public. The repository is now on GitHub at [mrabarnett/mrab-regex](https://github.com/mrabarnett/mrab-regex).

**Comment by facelessuser** [comment]
Good to know! Glad to see things move over to GitHub.

**Comment by mrabarnett** [comment]
Fixed in regex 2021.11.2.
